### ISSUE TITLE
Fish goes 100% CPU when I go into "less" mode in `history`

### ISSUE THREAD

**Issue by pasniak** [fish-shell/fish-shell#5438]
Fish goes 100% CPU (actually - of a core - which still sends the fan spinning) when I go int "vi/less mode" - for example when I execute `history` (I have ~400 lines)

fish, version 3.0.0
Linux Beast 4.15.0-43-generic #46-Ubuntu SMP Thu Dec 6 14:45:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
xterm-256color

In `sh -c 'env HOME=$(mktemp -d) fish'` things look ok, but maybe because there is no history?

This is my first try of fish.

**Comment by faho** [comment]
>when I go int "vi mode" 

I'm sorry, I don't understand this part. What commands did you execute?

**Comment by pasniak** [comment]
`history` 
![peek 2018-12-29 04-43](https://user-images.githubusercontent.com/19778666/50536639-5d2b8a80-0b4e-11e9-9f43-5e1e9aaa978c.gif)

**Comment by faho** [comment]
@pasniak: What does "go int vi mode" mean?

**Comment by pasniak** [comment]
When history of commands is long, `history` goes into different way to pad through the contents, which looks like `vi` or rather `less`-style panning, with colon at the bottom, '/' slash being search, `:100` going to 100th line etc. When I do `less` on some big file, I do not see the behavior.

**Comment by faho** [comment]
>rather less-style panning

That's because it _is_ `less`. We start $PAGER here.

One possibility is that your history seems to contain some lines that aren't actually history. If one of those e.g. was too long, and your less had issues with that, it would explain it.

So, please try:

```fish
history > file
less --quit-if-one-screen --no-init -c file
```

and then go into search mode. If that triggers it, you can figure out the longest line with

```fish
history | awk '{print length, $0}' | sort -nr | head -n 1
```

**Comment by pasniak** [comment]
nope, 
```
history > file
less --quit-if-one-screen --no-init -c file
```
does not trigger 100% core.

**Comment by faho** [comment]
And 

```fish
history | less --quit-if-one-screen --no-init -c
```

?

What is the value of $PAGER?

**Comment by pasniak** [comment]
`history | less --quit-if-one-screen --no-init -c` no joy. No 100% cpu.

```
michal@Beast /tmp> echo $PAGER

michal@Beast /tmp> 
```

**Comment by pasniak** [comment]
I started new `fish --debug-level 5`, this gives me continuous scroll off:
```
<5> fish: Skipping wait on incomplete job 1 (history ...)
<4> fish: select_try: no IO fds
<4> fish: Waiting on processes from foreground job 9779
<5> fish: Skipping wait on incomplete job 1 (history ...)
<4> fish: select_try: no IO fds
...
```

**Comment by faho** [comment]
Ah, okay, that explains it a bit.

@mqudsi: I get a loop of "select_try: no IO fds" with:

```fish
function t
    echo | less
end
t
```

Pressing "q" to quit less will solve it.

**Comment by mqudsi** [comment]
Thanks for pinning that down, @faho. It needs to be handled similar to the case when a command is backgrounded, although I'm not sure how that would be detected.

**Comment by faho** [comment]
@mqudsi: So... this is busy spinning because there are no fds it can read from in that job, right? Can that number ever increase once a job is running? If not, that's the case you'd have to detect.

**Comment by mqudsi** [comment]
@faho yes, normally it would sleep via `waitpid(2)` until the status of a process has changed. 

> Can that number ever increase once a job is running?

What number are you referring to?

**Comment by faho** [comment]
>What number are you referring to?

The number of fds to read from. If select_try ever returns `IOCHAIN_EMPTY` for a job, doesn't every subsequent call also return that, for that specific job?

**Comment by mqudsi** [comment]
Presumably. (It's hard to answer definitively about job control!)

It's not a question of finding an integral state to monitor but rather tying this to the lifetime of a process, as what we need to do here is to block until a process terminates (or is suspended). 

The problem isn't with the general case, which appears to be handled both correctly and well (which makes sense, since "fish spins in a tight loop if IO is being chained from one process to another" is rather a rather extreme oversight). e.g. run `yes | less` in a terminal and monitor the parent `fish` process CPU usage in another (it's the highlighted process below):

![image](https://user-images.githubusercontent.com/606923/50552590-3ce9f180-0c5c-11e9-8fb0-712bd479f865.png)

And it's not even a case of "fish builtin piping to a pager," as running `echo | less` directly doesn't have any problems, either. (These cases all end up with the same fd count situation, fwiw). I thought it was the exemption from ever waiting on our own pid that was causing this, but since executing both your `t` and `echo | less` don't spawn a new process for the `echo` builtin, that's clearly not the case, either.

This may be related to the changes pertaining to recursive job invocations. When `t` is started, you have job 1 running `t` and then a new job is spawned to execute the contents of the `t` function. Here's what `jobs` emits after `t` is suspended via `ctrl-z`:

```
mqudsi@ZBOOK /m/c/U/mqudsi> jobs
Job     Group   CPU     State   Command
2       12394   0%      stopped echo | less
```

The real job that's executing `t` is supressed from that list, but it's job 1. (Side note: this is a bug, if a job were to be listed, you'd think it would be the top-most `t` and not the inner `echo | less`. Second side note: we should add a flag that would list all jobs, including the hidden ones).

**Comment by mqudsi** [comment]
We are actually waiting on the process correctly.

I added this bit to the `waitpid` calls:
![image](https://user-images.githubusercontent.com/606923/50552930-fea40080-0c62-11e9-9c87-6244026f393e.png)

Which produced the following output:
```
<4> fish: Waiting on processes from foreground job 1553
<4> fish: Waiting on individual process 1553
<4> fish: select_try: no IO fds
```

But that PID is incorrect!

![image](https://user-images.githubusercontent.com/606923/50552940-21ceb000-0c63-11e9-90f9-22f4704dced8.png)

That's the pgrp and not the pid, but it's stored as the pid.

**Comment by mqudsi** [comment]
This can be fast-tracked for fish 3.0.1

**Comment by ridiculousfish** [comment]
In the case of `echo | less`, we fork to write out echo's (single-byte) output. This fork adds a process to the job's process list. This `echo` then exits and the process is marked as complete.

In `process_mark_finished_children` we wait on the exited `echo` process (bad, the pid could have been recycled). This produces `ECHILD`. We then set WNOHANG (because we should only set WNOHANG once) and `waitpid()` on `less`, which is not finished, so it returns immediately.

Fixes:
1. Don't wait on finished processes.
2. Only set WNOHANG if we actually got a pid back.

Fix the first went into master in 2ebdcf82eebec230549b64dc740ea604c7772049, I want to have both in 3.0.1.

**Comment by mqudsi** [comment]
At this point it’s necessary for the waitpid call to return a valid pid since the process in question hasn’t exited. If a process isn’t waited on without WNOHANG this time around, it will be next time. 

The problem is that in the released 3.0.0 code, the same process (after being reaped) was being waited on in a blocking fashion, but that’s no longer the case.

**Comment by ridiculousfish** [comment]
This went into 3.0.1 as e2f2dbf032c0a3bcc13b91a5d56cec2f5e294027
